### DISCUSSION TITLE
Graceful shutdown of websockets

### DISCUSSION THREAD

**Discussion by aviramha** [Kludex/uvicorn#1137] [Potential Issue]
Hi!
We've encountered a situation where Kubernetes terminates a pod, and Websocket connections aren't gracefully shutdown.
I debugged it a bit, and found out how Uvicorn behaves on termination - on simple HTTP - it just cancels the keep alive then awaits the server tasks.
On the websocket, it closes the transport on currently "running" websockets.
I want to suggest changing this behavior to only stop serving new connections, without shutting down the connections. 
The code change would be just removing the `transport.close` on each websocket implementations, leaving the server.py code to await pending tasks, and canceling those if needed.

Let me know if I can send a PR to apply this change.

**Comment by euri10** [comment]
iirc I added this while working on making pytest stricter https://github.com/encode/uvicorn/pull/929 and specifically the ws addition you mention I was unsure about, see this [comment](https://github.com/encode/uvicorn/pull/929#discussion_r582803783) for further context.

so I'd be ok seeing a PR that removes that, but we just need to be sure that :
1. it doesn't pop unrelated ResourceWarning raised by pytest, 
2. that the fix would not be better handled by https://github.com/encode/uvicorn/pull/853 or https://github.com/encode/uvicorn/pull/1069 if this is the case (several processes) because I'm not too sure if those kind of issues are not caused more generally by our long-standing "graceful shutdown issue"
